perception/autoware_tensorrt_bevdet/src/nvjpegdecoder.cpp

```diff
@@ -0,0 +1,299 @@
+
+  CHECK_CUDA(cudaMemcpy2D(
+    img + 0 * step, static_cast<size_t>(width), d_chanR, static_cast<size_t>(pitchR), width, height,
+    cudaMemcpyDeviceToDevice));
+  CHECK_CUDA(cudaMemcpy2D(
+    img + 1 * step, static_cast<size_t>(width), d_chanG, static_cast<size_t>(pitchG), width, height,
+    cudaMemcpyDeviceToDevice));
+  CHECK_CUDA(cudaMemcpy2D(
+    img + 2 * step, static_cast<size_t>(width), d_chanB, static_cast<size_t>(pitchB), width, height,
+    cudaMemcpyDeviceToDevice));
+
+  return EXIT_SUCCESS;
+}
+
+void decode_single_image(
+  const std::vector<char> & img_data, nvjpegImage_t & out, decode_params_t & params,
+  share_params & share_param, double & time)
+{
+  CHECK_CUDA(cudaStreamSynchronize(params.stream));
+  cudaEvent_t startEvent = NULL, stopEvent = NULL;
+  float loopTime = 0;
+  // cudaEventBlockingSync
+  CHECK_CUDA(cudaEventCreate(&startEvent));
+  CHECK_CUDA(cudaEventCreate(&stopEvent));
+
+  CHECK_CUDA(cudaEventRecord(startEvent, params.stream));
+
+  CHECK_NVJPEG(nvjpegStateAttachDeviceBuffer(params.nvjpeg_decoupled_state, params.device_buffer));
+  int buffer_index = 0;
+  CHECK_NVJPEG(nvjpegDecodeParamsSetOutputFormat(params.nvjpeg_decode_params, share_param.fmt));
+
+  CHECK_NVJPEG(nvjpegJpegStreamParse(
+    share_param.nvjpeg_handle, reinterpret_cast<const uchar *>(img_data.data()), img_data.size(), 0,
+    0, params.jpeg_streams[buffer_index]));
+
+  CHECK_NVJPEG(nvjpegStateAttachPinnedBuffer(
+    params.nvjpeg_decoupled_state, params.pinned_buffers[buffer_index]));
+
+  CHECK_NVJPEG(nvjpegDecodeJpegHost(
+    share_param.nvjpeg_handle, params.nvjpeg_decoder, params.nvjpeg_decoupled_state,
+    params.nvjpeg_decode_params, params.jpeg_streams[buffer_index]));
+
+  CHECK_CUDA(cudaStreamSynchronize(params.stream));
+
+  CHECK_NVJPEG(nvjpegDecodeJpegTransferToDevice(
+    share_param.nvjpeg_handle, params.nvjpeg_decoder, params.nvjpeg_decoupled_state,
+    params.jpeg_streams[buffer_index], params.stream));
+
+  buffer_index = 1 - buffer_index;  // switch pinned buffer in pipeline mode to avoid an extra sync
+  RCLCPP_DEBUG(rclcpp::get_logger("nvjpegDecoder"), buffer_index);
+
+  CHECK_NVJPEG(nvjpegDecodeJpegDevice(
+    share_param.nvjpeg_handle, params.nvjpeg_decoder, params.nvjpeg_decoupled_state, &out,
+    params.stream));
+
+  // CHECK_CUDA(cudaStreamSynchronize(params.stream)); //TODO new add
+
+  CHECK_CUDA(cudaEventRecord(stopEvent, params.stream));
+
+  CHECK_CUDA(cudaEventSynchronize(stopEvent));
+  CHECK_CUDA(cudaEventElapsedTime(&loopTime, startEvent, stopEvent));
+  time = 0.001 * static_cast<double>(loopTime);  // cudaEventElapsedTime returns milliseconds
+                                                 // time = 1.0;
+}
+
+nvjpegDecoder::nvjpegDecoder(size_t n, size_t _fmt)
+: N_img(n), iout(n), isz(n), widths(n), heights(n), params(n), share_param(_fmt)
+{
+  init();
+}
+
+int nvjpegDecoder::decode(const std::vector<std::vector<char>> & files_data, uchar * out_imgs)
```

Comment:
The `decode` function uses `std::thread` for parallel decoding. While this improves performance, it's crucial to ensure proper exception handling within the threads.  Consider using a more robust approach like `std::future` or a thread pool library, which provide better mechanisms for managing and handling exceptions from parallel threads.